src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -418,117 +255,56 @@ private Type EmitCoercionCode(AssociativeNode arg, Type argType, Type param)
-
-            // i = 0;
-            localBuilder = DeclareLocal(typeof(int), "for loop counter");
-            var counterIndex = localBuilder.LocalIndex;
-            EmitOpCode(OpCodes.Ldc_I4_0);
-            EmitOpCode(OpCodes.Stloc, counterIndex);
-
-            var loopBodyLabel = DefineLabel();
-            var loopCondLabel = DefineLabel();
-
-            EmitOpCode(OpCodes.Br_S, loopCondLabel.Value);
 
-            // newarr[i] = (Target)arr[i];
-            MarkLabel(loopBodyLabel.Value, "label:body");
-
-            EmitOpCode(OpCodes.Ldloc, newArrIndex);
-            EmitOpCode(OpCodes.Ldloc, counterIndex);
-
-            
-            EmitOpCode(OpCodes.Ldloc, currentVarIndex);
-            EmitOpCode(OpCodes.Ldloc, counterIndex);
-
-            if (typeof(Source) == typeof(double))
-            {
-                EmitOpCode(OpCodes.Ldelem_R8);
-            }
-            else if (typeof(Source) == typeof(long))
-            {
-                EmitOpCode(OpCodes.Ldelem_I8);
-            }
-            else if (typeof(Source) == typeof(int))
+            EmitOpCode(OpCodes.Ldnull);
+            MethodInfo convertMethod = null;
+            if (typeof(Target) == typeof(int))
             {
-                EmitOpCode(OpCodes.Ldelem_I4);
+                convertMethod = typeof(Convert).GetMethod(nameof(Convert.ToInt32), new[] { typeof(Source) });
             }
-            if (typeof(Target) == typeof(int))
+            else if(typeof(Target) == typeof(double))
             {
-                if (typeof(Source) == typeof(double))
-                {
-                    EmitMathRound();
-                }
-                EmitOpCode(OpCodes.Conv_I4);
-                EmitOpCode(OpCodes.Stelem_I4);
+                convertMethod = typeof(Convert).GetMethod(nameof(Convert.ToDouble), new[] { typeof(Source) });
             }
             else if (typeof(Target) == typeof(long))
             {
-                if (typeof(Source) == typeof(double))
-                {
-                    EmitMathRound();
-                }
-                EmitOpCode(OpCodes.Conv_I8);
-                EmitOpCode(OpCodes.Stelem_I8);
+                convertMethod = typeof(Convert).GetMethod(nameof(Convert.ToInt64), new[] { typeof(Source) });
             }
-            if (typeof(Target) == typeof(double))
+            else if (typeof(Target) == typeof(bool))
             {
-                EmitOpCode(OpCodes.Conv_R8);
-                EmitOpCode(OpCodes.Stelem_R8);
+                convertMethod = typeof(Convert).GetMethod(nameof(Convert.ToBoolean), new[] { typeof(Source) });
             }
-            // i++;
-            EmitOpCode(OpCodes.Ldloc, counterIndex);
-            EmitOpCode(OpCodes.Ldc_I4_1);
+            EmitOpCode(OpCodes.Ldftn, convertMethod);
 
-            EmitOpCode(OpCodes.Add);
+            var ctor = System.Linq.Expressions.Expression.GetDelegateType(typeof(Source), typeof(Target))
```

Comment:
so this is creating a delegate that returns the target type and takes the source?
I see that this requires referencing codeDom... and linq expression trees - do you think this additional dependency will cause issues for us? Is it supported in .net 6/7/8?

Reply:
This doesn't require any special dependencies. I just removed the `using System.Codedom` as it wasn't referenced anywhere in this file.